restorer.go

```diff
@@ -0,0 +1,100 @@
+package lifecycle
+
+import (
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/pkg/errors"
+
+	"github.com/buildpack/lifecycle/archive"
+	"github.com/buildpack/lifecycle/cmd"
+	"github.com/buildpack/lifecycle/image"
+)
+
+type Restorer struct {
+	LayersDir  string
+	Buildpacks []*Buildpack
+	Out, Err   *log.Logger
+}
+
+func (r *Restorer) Restore(cacheImage image.Image) error {
+	if found, err := cacheImage.Found(); !found || err != nil {
+		r.Out.Printf("cache image '%s' not found, nothing to restore", cacheImage.Name())
+		return nil
+	}
+	metadata, err := getCacheMetadata(cacheImage, r.Out)
+	if err != nil {
+		return err
+	}
+	for _, bp := range r.Buildpacks {
+		layersDir, err := readBuildpackLayersDir(r.LayersDir, *bp)
+		if err != nil {
+			return err
+		}
+		bpMD := metadata.metadataForBuildpack(bp.ID)
+		for name, layer := range bpMD.Layers {
+			if !layer.Cache {
+				continue
+			}
+
+			bpLayer := layersDir.newBPLayer(name)
+			r.Out.Printf("restoring cached layer '%s'", bpLayer.Identifier())
+			if err := bpLayer.writeMetadata(bpMD.Layers); err != nil {
+				return err
+			}
+			if layer.Launch {
+				if err := bpLayer.writeSha(layer.SHA); err != nil {
+					return err
+				}
+			}
+			rc, err := cacheImage.GetLayer(layer.SHA)
+			if err != nil {
+				return err
+			}
+			defer rc.Close()
```

Comment:
This will close all of the layers at the end of the whole method, not at the end of each iteration. This will leak file descriptors. I suggest keeping the defer, but extracting a method here.